src/runtime-rs/crates/resource/src/share_fs/sandbox_bind_mounts.rs

```diff
@@ -0,0 +1,164 @@
+        let host_mounts_path = PathBuf::from(bindmounts_path);
+
+        Ok(SandboxBindMounts {
+            sid,
+            host_mounts_path,
+            sandbox_bindmounts,
+        })
+    }
+
+    pub fn bind_mounts_empty(&self) -> bool {
+        if self.sandbox_bindmounts.is_empty() {
+            info!(sl!(), "sandbox bind mounts empty, skip it.");
+            return true;
+        }
+
+        false
+    }
+
+    fn parse_sandbox_bind_mounts<'a>(&self, bindmnt_src: &'a str) -> Result<(&'a str, &'a str)> {
+        // get the bindmount's r/w mode
+        let bindmount_rw = if bindmnt_src.ends_with(SANDBOX_BIND_MOUNTS_RW) {
+            SANDBOX_BIND_MOUNTS_RW
+        } else {
+            SANDBOX_BIND_MOUNTS_RO
+        };
+
+        // get the true bindmount from the string
+        let bindmount = bindmnt_src.trim_end_matches(bindmount_rw);
+
+        Ok((bindmount_rw, bindmount))
+    }
+
+    pub async fn setup_sandbox_bind_mounts(&self) -> Result<()> {
+        let mut mounted_list: Vec<PathBuf> = Vec::new();
+        let mut mounted_map: HashMap<String, bool> = HashMap::new();
+        for src in &self.sandbox_bindmounts {
+            let (bindmount_rw, bindmount) = self
+                .parse_sandbox_bind_mounts(src)
+                .context("parse sandbox bind mounts failed")?;
+
+            // get the basename of the canonicalized mount path mnt_name: dirX
+            let mnt_name = get_base_name(bindmount)?
+                .into_string()
+                .map_err(|e| anyhow!("failed to get base name {:?}", e))?;
+
+            // if repeated mounted, do umount it and return error
+            if mounted_map.insert(mnt_name.clone(), true).is_some() {
+                for p in &mounted_list {
+                    let _ = nix::mount::umount(p);
+                }
+
+                return Err(anyhow!(
+                    "sandbox-bindmounts: path {} is already specified.",
+                    bindmount
+                ));
+            }
+
+            // mount_dest: /run/kata-containers/shared/sandboxes/<sid>/rw/passthrough/sandbox-mounts/dirX
+            let mount_dest = self.host_mounts_path.clone().join(mnt_name.as_str());
+            mkdir_with_permissions(self.host_mounts_path.clone().to_path_buf(), 0o750).context(
+                format!(
+                    "create host mounts path {:?}",
+                    self.host_mounts_path.clone()
+                ),
+            )?;
+
+            info!(
+                sl!(),
+                "sandbox-bindmounts mount_src: {:?} => mount_dest: {:?}", bindmount, &mount_dest
+            );
+
+            // mount -o bind,ro host_shared mount_dest
+            // host_shared: ${bindmount}
+            mount::bind_mount_unchecked(Path::new(bindmount), &mount_dest, true).map_err(|e| {
+                for p in &mounted_list {
+                    let _ = nix::mount::umount(p);
+                }
+                e
```

Comment:
Please add some explanation with the error.

Reply:
fixed